fdbclient/S3Cp.actor.cpp

```
@@ -0,0 +1,579 @@
+			return false;
+		}
+	}
+	if (tlsVerifyPeers.size()) {
+		try {
+			setNetworkOption(FDBNetworkOptions::TLS_VERIFY_PEERS, tlsVerifyPeers);
+		} catch (Error& e) {
+			std::cerr << "ERROR: cannot set TLS peer verification to " << tlsVerifyPeers << " (" << e.what() << ")\n";
+			return false;
+		}
+	}
+	return true;
+}
+
+// Get the endpoint for the given s3url.
+// Populates parameters and resource with parse of s3url.
+static Reference<S3BlobStoreEndpoint> getEndpoint(std::string s3url,
+                                                  std::string& resource,
+                                                  S3BlobStoreEndpoint::ParametersT& parameters) {
+	std::string error;
+	Reference<S3BlobStoreEndpoint> endpoint =
+	    S3BlobStoreEndpoint::fromString(s3url, {}, &resource, &error, &parameters);
+	if (resource.empty()) {
+		TraceEvent(SevError, "EmptyResource").detail("s3url", s3url);
+		throw backup_invalid_url();
+	}
+	for (auto c : resource) {
+		if (!isalnum(c) && c != '_' && c != '-' && c != '.' && c != '/') {
+			TraceEvent(SevError, "IllegalCharacter").detail("s3url", s3url);
+			throw backup_invalid_url();
+		}
+	}
+	if (error.size()) {
+		TraceEvent(SevError, "GetEndpointError").detail("s3url", s3url).detail("error", error);
+		throw backup_invalid_url();
+	}
+	return endpoint;
+}
+
+// Copy filepath to bucket at resource in s3.
+ACTOR Future<Void> copy_up_file(Reference<S3BlobStoreEndpoint> endpoint,
+                                std::string bucket,
+                                std::string resource,
+                                std::string filepath) {
+	// Reading an SST file fully into memory is pretty obnoxious. They are about 16MB on
+	// average. Streaming would require changing this s3blobstore interface.
+	// Make 32MB the max size for now even though its arbitrary and way to big.
+	state std::string content = readFileBytes(filepath, 1024 * 1024 * 32);
```

what happens if file is larger than 32MB?

---

readFileBytes will throw file_too_large exception
```

	size_t size = fileSize(filename);
	if (size > maxSize) {
		TraceEvent("ReadFileBytes_FileTooLarge")
		    .detail("FileSize", size)
		    .detail("InputMaxSize", maxSize)
		    .detail("Filename", filename);
		throw file_too_large();
	}

```

---

(Thanks for reviewing) Would you like a different behavior?

---

That should be good for testing. I just want to make sure we are not silently masking errors.